main.rb

```diff
@@ -0,0 +1,98 @@
+module Enumerable
+  def my_each
+    return to_enum(:my_each) unless block_given?
+
+    size.times { |i| yield(to_a[i]) }
+    self
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    size.times do |i|
+      yield to_a[i], i
+    end
+    self
+  end
+
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    ary = []
+    my_each do |elem|
+      ary << elem if yield elem
+    end
+    ary
+  end
+
+  def my_all?
+    return nil unless block_given?
+
+    my_each do |elem|
+      return false unless yield elem
+    end
+    true
+  end
+
+  def my_map(proc = nil)
+    return dup unless block_given?
+    return dup if proc.nil?
+
+    ary = []
+    if proc.nil?
+      to_a.my_each { |elem| ary << yield(elem) }
+    else
+      to_a.my_each { |elem| ary << proc.call(elem) }
+    end
+    ary
+  end
```

Comment:
- enumerables #my_map returns a new array with the results of running a given block once for every element in enum.
- enumerables #my_map returns a new array with the results of running a given block once for every element in enum.
- enumerables #my_map returns a new array with the results of running a given block once for every element in enum::range
- enumerables #my_map returns an Enumerator if no block is given
- enumerables #my_map accepts a proc as an argument and returns a new array with the results of running the proc once for every element in enum.